### ISSUE TITLE
Infinite loop on reauth process

### ISSUE THREAD

**Issue by garbage-collector** [gophercloud/gophercloud#1182]
I was performing some tests on gophercloud when I encountered an issue.

Let's imagine that compute always give a 401 answer but keystone is always answering 201 on token creation.

This situation ends with an infinite loop on reauthentication process.

You can easily reproduce the issue with this code (I'm using gock to fake the OpenStack API):

```golang
package main

import (
	"log"

	"github.com/gophercloud/gophercloud"
	"github.com/gophercloud/gophercloud/openstack"
	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
	"github.com/gophercloud/gophercloud/openstack/identity/v3/tokens/testing"
	"github.com/h2non/gock"
)

func main() {
	gock.New("http://127.0.0.1:8774/v2.1/a99e9b4e620e4db09a2dfb6e42a01e66").
		Get("/servers/detail").
		MatchHeader("X-Auth-Token", "TOKEN").
		Persist().
		Reply(401)

	gock.New("http://127.0.0.1:5000/v3/").
		Post("/auth/token").
		MatchType("json").
		Persist().
		Reply(201).
		SetHeader("X-Subject-Token", "TOKEN").
		JSON(testing.TokenOutput)

	opts := gophercloud.AuthOptions{
		IdentityEndpoint: "http://127.0.0.1:5000/v3/",
		Username:         "{username}",
		Password:         "{password}",
		DomainID:         "{domain-id}",
		AllowReauth:      true,
	}

	provider, err := openstack.AuthenticatedClient(opts)
	if err != nil {
		log.Fatal(err)
	}

	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
		Region: "RegionOne",
	})
	if err != nil {
		log.Fatal(err)
	}

	pages, err := servers.List(client, nil).AllPages()
	if err == nil {
		log.Fatal("authentication required expected")
	}
	if pages != nil {
		log.Fatal("pages expected to be nil")
	}
}
```

This code will have an infinite loop, without any logs or print, so don't forget to stop the program.

After digging the code, the only behaviour handled by the reauthentication are those two:
* Compute returns HTTP 401
* Keystone returns HTTP 201
* Compute returns HTTP 200

or

* Compute returns HTTP 401
* Keystone returns HTTP 401

This scenario does not seems to be handled:
* Compute returns HTTP 401
* Keystone returns HTTP 201
* Compute returns HTTP 401

Currently looking to propose a PR.

**Comment by garbage-collector** [comment]
PR created: https://github.com/gophercloud/gophercloud/pull/1192

**Comment by jtopjian** [comment]
Fixed in #1192
